### PR TITLE
WDPD-42: Giropay payment method

### PR DESCRIPTION
**Pull request by ps-holler**
### This PR

* Adds Giropay payment method

### Notes

* No backend operations are available for Giropay at this time (see https://github.com/wirecard/magento2-ee/wiki/giropay)

### How to Test

* Open the admin, go to _Extensions → Modules_ and activate the module
* Go to _Shop Settings → Payment Methods_ and activate _Wirecard Giropay_
* Open the shop and make a purchase using Giropay as a payment method
* For the BIC, use a BIC of [one of the supporting banks](https://www.giropay.de/kaeufer/) (e.g. `BYLADEM1001`, `BYLADEM1KIS`)
* Abort the payment and use the [Test Bank Account BIC](https://document-center.wirecard.com/display/PTD/giropay#giropay-TestCredentials)
* In the test account screen, trigger a successful payment by entering these values:
  * sc: `10`
  * extensionSc: `4000`
* Confirm that the payment was successful by going to Wirecard _Transactions → Transactions_ and _Administer Orders → Orders_ in the admin

### Jira Links

* WDPD-42: https://jira.parkside.at/browse/WDPD-42

### CODE REVIEW

File: Core/PaymentMethodFactory.php [hunk 1]
```diff
@@ -35,28 +55,16 @@ class PaymentMethodFactory
...
-                return new CreditCardPaymentMethod();
-            case SepaCreditTransferPaymentMethod::getName(true):
-                return new SepaCreditTransferPaymentMethod();
-            case SepaDirectDebitPaymentMethod::getName(true):
-                return new SepaDirectDebitPaymentMethod();
-            case SofortPaymentMethod::getName(true):
-                return new SofortPaymentMethod();
-            default:
-                throw new StandardException("payment type not registered: $sPaymentMethodType");
+        foreach (self::getPaymentMethodClasses() as $sClassName) {
```
Comment (ps-cgrach) [inline Core/PaymentMethodFactory.php]: Nice! 👍

File: Tests/Unit/Model/GiropayPaymentMethodTest.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        $this->assertEquals(array_keys($aFields), [
+            'bic',
+        ]);
+    }
+
+    public function testGetPublicFieldNames()
+    {
+        $aFieldNames = $this->_oPaymentMethod->getPublicFieldNames();
+
+        $this->assertEquals($aFieldNames, [
```
Comment (ps-cgrach) [inline Tests/Unit/Model/GiropayPaymentMethodTest.php]: same for the parameters as above

File: default_payment_config.xml [hunk 3]
```diff
@@ -29,6 +29,8 @@
         <wdoxidee_httppass>qD2wzQ_hrc!8</wdoxidee_httppass>
         <wdoxidee_descriptor>1</wdoxidee_descriptor>
         <wdoxidee_additional_info>1</wdoxidee_additional_info>
+        <wdoxidee_delete_canceled_order>0</wdoxidee_delete_canceled_order>
```
Comment (ps-cgrach) [inline default_payment_config.xml]: I think default values are not needed to be defined. 🤔

File: Model/GiropayPaymentMethod.php [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    /**
+     * @inheritdoc
+     *
+     * @param string $sAction
+     *
+     * @return SepaCreditTransferPaymentMethod
+     *
+     * @since 1.2.0
+     */
+    public function getPostProcessingPaymentMethod($sAction)
```
Comment (ps-amehic) [inline Model/GiropayPaymentMethod.php]: Unused parameter

File: Model/GiropayPaymentMethod.php [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+     */
+    public function getTransaction()
+    {
+        $oTransaction = new GiropayTransaction();
+        $oBankAccount = new BankAccount();
+        $oSession = Registry::getConfig()->getSession();
+        $aDynvalues = $oSession->getVariable('dynvalue');
+
+        $oBankAccount->setBic($aDynvalues['bic'] ?? '');
+        $oTransaction->setBankAccount($oBankAccount);
```
Comment (ps-amaureder) [inline Model/GiropayPaymentMethod.php]: This should be set in addMandatoryTransactionData()

File: Core/PaymentMethodFactory.php [hunk 6]
```diff
@@ -27,6 +28,25 @@
...
+     * Returns an array of payment method class names.
+     *
+     * @return array
+     *
+     * @since 1.2.0
+     */
+    public static function getPaymentMethodClasses()
+    {
+        return [
+            PaypalPaymentMethod::class,
```
Comment (ps-dfritzsch) [inline Core/PaymentMethodFactory.php]: Is the order of payment methods relevant?

If not, let's please sort it **alphabetically** instead (except Credit Card that is the first).

**Pull request merged**